### ISSUE TITLE
DebugName memory leak

### ISSUE THREAD

**Issue by OndrejPetrzilka** [sharpdx/SharpDX#850]
There's memory leak when setting DebugName of the object.

```
public string DebugName
{
    get { ... }
    set
    {
        if (string.IsNullOrEmpty(value))
        {
            SetPrivateData(CommonGuid.DebugObjectName, 0, IntPtr.Zero);
        }
        else
        {
            var namePtr = Utilities.StringToHGlobalAnsi(value);
            SetPrivateData(CommonGuid.DebugObjectName, value.Length, namePtr);
        }
    }
}
```
**StringToHGlobalAnsi** creates copy of string in unmanaged memory.
**SetPrivateData** makes a copy of the specified data and stores it with the object, see [here](https://msdn.microsoft.com/en-us/library/windows/desktop/bb174544(v=vs.85).aspx)
What's missing is **Marshal.FreeHGlobal** of that native string.

Source:
https://github.com/sharpdx/SharpDX/blob/8dc1f64a9424c6b0dc8b9009f70e144ded03f1d7/Source/SharpDX.DXGI/DeviceChild.cs#L71

**EDIT:**
Even better than Marshal.FreeHGlobal would be to stackalloc memory and convert string into this memory (ANSI string is needed).

**Comment by xoofx** [comment]
This is a known issue. In the past, when I tried to deallocate the string right after `SetPrivateData`, it was crashing, because somehow `SetPrivateData` didn't seem to copy the string but kept the pointer and this could give some weird errors later when trying to debug DX. I don't know if this still holds, but at least you are aware of some underlying pitfalls...

**Comment by xoofx** [comment]
This is even older than I thought done by commit bf02736ba8ba5d36d21e95fa4251307f77082b4c

**Comment by OndrejPetrzilka** [comment]
Memory was allocated with **Marshal.StringToHGlobalAnsi**, and it was released with **Marshal.Release**.

That's probably the issue, **Marshal.Release** is designed for COM objects (and bad things happen when it tries to call IUnknown.Release on string pointer :))

Releasing it with **Marshal.FreeHGlobal** works fine for me.

**Comment by xoofx** [comment]
oops yes, good catch oO don't know why I was using Marshal.Release in the first place.... so PR welcome to fix this.

**Comment by ab4d** [comment]
Because there is no recent activity on this issue, I wanted to provide a pool request to fix this.

But on my surprise I found a few problems when trying to provide the best fix.

As OndrejPetrzilka recommended in his edit, I also wanted to fix this without using alloc (used by Marshal.StringToHGlobalAnsi).

At first I wanted to follow OndrejPetrzilka advice and use stackalloc. But I did not find an efficient way to convert string to ANSI and copy it into stackalloc. If we would have access to the string's ConvertToAnsi method (used by Marshal.StringToHGlobalAnsi), this would be possible. But this method is internal. So we would first need to convert string to ANSI byte array, then get pointer to this array and copy the array to stackalloc. But once we have the pointer to ANSI byte array, we can pass it to the SetPrivateData without need to use stackalloc.

Therefore I ended up with the following code:

```
// Convert to ANSI byte array (using current OS code page)
byte[] nameBytes = System.Text.Encoding.Default.GetBytes(value); 

unsafe
{
    fixed (byte* namePtr = nameBytes)
    {
        // MSDN: SetPrivateData allocates an internal buffer for the data and copies it. You can then safely free the source buffer or object.
        SetPrivateData(CommonGuid.DebugObjectName, nameBytes.Length, new IntPtr(namePtr));
    }
}
```

But after I put that code into the SharpDX source code, I found out that System.Text.Encoding.Default does not exist in the portable .Net. What is more, it looks like there is no support for ANSI and even ASCII in portable .Net.

An option would be to use the ASCII encoding from SharpDX:
byte[] nameBytes = SharpDX.Text.ASCIIEncoding.ASCII.GetBytes(value);


But this would prevent using the characters that are special for the current ANSI code page and that are currently preserved.

So at the end, it looks like using StringToHGlobalAnsi method is the only option here and that this issue can be fixed with simply adding Marshal.FreeHGlobal (also because DebugName is not performance critical).

Or maybe I am missing something. What do you think?

**Comment by OndrejPetrzilka** [comment]
Can you use this overload?
[Encoding.GetBytes Method (Char*, Int32, Byte*, Int32)](https://msdn.microsoft.com/en-us/library/6s1x2atd(v=vs.110).aspx)

There's also this method:
[Encoding.GetMaxByteCount Method (Int32)](https://msdn.microsoft.com/en-us/library/system.text.encoding.getmaxbytecount(v=vs.110).aspx)

It would work something like this:
1) int len = GetMaxByteCount(value.Length)
2) byte* ansi = stackalloc byte[len]
3) fixed(char* managed = value)
4) GetBytes(managed, ..., ansi, ...)
5) SetPrivateData(..., ansi, ...)

In case it's not possible without allocations, **Marshal.StringToHGlobalAnsi** seems preferable (at least it does not create garbage for GC)

**Comment by xoofx** [comment]
Fixed by #854
